### DISCUSSION TITLE
Change of encoding backend

### DISCUSSION THREAD

**Discussion by MimoKar** [bigcat88/pillow_heif#309] [Q&A]
Is there a possibility to change the encoding backend for AVIF from AOM to svt-av1 f.e?

**Comment by bigcat88** [comment]
Yes, it is supported. If you build `pillow-heif` from source and `libheif` has that backend enabled you can set preferred backend with these:

https://github.com/bigcat88/pillow_heif/blob/03c7bebae111041e1d4de1db7741b5d506d9c8f0/pillow_heif/options.py#L68-L91

here is even a test for it:

https://github.com/bigcat88/pillow_heif/blob/03c7bebae111041e1d4de1db7741b5d506d9c8f0/tests/write_test.py#L626-L641

for Pillow plugin alternative way to set option is to pass `preferred_encoder` as an argument to `register_avif_opener` or `register_heif_opener`

**Comment by MimoKar** [comment]
Ok. As it appears that I am dumb to compile it myself (spent a week on it before)  is it possible to include the additional backends to the default package?

**Reply by bigcat88** [reply]
Unfortunately **no**, the package size can easily grow to 50+ MB for each wheel, and space on `pypi` is limited for each package.

If you are using Ubuntu, it is easy to install using `PPA` with libheif releases.
Another options is to create a discussion or issue in the `libheif` repository, and there will help you with building libheif with all these backends.

After that, building pillow-heif will be very easy, there should be no problems.

**Comment by MimoKar** [comment]
Don't plan to change backend to quicker one - SVT-AV1?

**Comment by MimoKar** [comment]
When I want to use a plugin system in libheif, where I need to put a DLL file for a specific codec that it will be used by pillow_heif?

**Reply by bigcat88** [reply]
if `libehif` is built with the PluginSystem enabled, than the value of this variable determines the directory from which load plugins:

https://github.com/strukturag/libheif/blob/2a764da6053f08b472f7127f8d95772e64ef015a/CMakeLists.txt#L78-L79

**Reply by MimoKar** [reply]
```
Is your libheif used in pillow-HEIF built with a plugin system? 
If not, it is possible to do it?
```

Plugin system is available on Linux only :-(

**Comment by MimoKar** [comment]
Ok, I was able to build libheif in mingw.
But now pillow_heif build is throwing this error:

```
Found 'heif.h' in 'D:\gitrepos\pillow_heif\libheif'
building '_pillow_heif' extension
b'T\x00h\x00e\x00 \x00s\x00y\x00n\x00t\x00a\x00x\x00 \x00o\x00f\x00 \x00t\x00h\x00e\x00 \x00c\x00o\x00m\x00m\x00a\x00n\x00d\x00 \x00i\x00s\x00 \x00i\x00n\x00c\x00o\x00r\x00r\x00e\x00c\x00t\x00.\x00\r\x00\n\x00'
error: Error executing cmd /u /c "C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Auxiliary\Build\vcvarsall.bat" x86_amd64 && set
```

**Reply by bigcat88** [reply]
it is the encoded in the unicode next message: `"The syntax of the command is incorrect.\r\n"`

but I have no clue what does it mean

**Reply by bigcat88** [reply]
I will think about bundling `svt-a1` into next release, maybe I will.
Do you need it only on Windows?

**Reply by MimoKar** [reply]
Thanks.
I am using pillow_heif only on Windows now.

**Reply by bigcat88** [reply]
I asked here if it would be possible to use AVIF from `pillow_heif` if they merged the PR from `pillow_avif_plugin`, which would add AVIF support to Pillow itself:

https://github.com/python-pillow/Pillow/issues/7983
